docs/specs/balance-fetcher.md

```diff
@@ -0,0 +1,311 @@
+   - For each missing date:
+     - Call `eth_getBalance` with distributor address and block number
+     - Convert balance from hex to decimal string
+     - Store in balance map
+   - Merge new balances with existing data
+   - Write updated balance file atomically
+
+### Error Handling
+
+- RPC failures trigger exponential backoff retry (up to 3 attempts)
+- Individual distributor failures don't stop processing of others
+- Failed fetches are logged and reported in result summary
+- Partial data is never written - updates are atomic
+
+## 7. Data Structures
+
+### Input Types
+
+```typescript
+interface FetchOptions {
+  distributors?: string[]; // Optional list of specific distributors to process
+  dates?: string[]; // Optional list of specific dates to process
+  forceRefresh?: boolean; // Force re-fetch even if data exists
+}
+
+interface DistributorInfo {
+  address: string;
+  metadata: {
+    name: string;
+    created_at: string; // ISO date when distributor was created
+  };
+}
+
+interface BlockNumbers {
+  [date: string]: number; // Map of ISO date to block number
+}
+```
+
+### Output Types
+
+```typescript
+interface FetchResult {
+  success: boolean;
+  distributorsProcessed: number;
+  balancesFetched: number;
+  errors: FetchError[];
+}
+
+interface FetchError {
+  distributor: string;
+  date: string;
+  error: string;
+}
+
+interface BalanceData {
+  metadata: {
+    chain_id: number;
+    reward_distributor: string;
+  };
+  balances: {
+    [date: string]: {
+      block_number: number;
+      balance_wei: string; // Decimal string to avoid precision loss
+    };
+  };
+}
+```
+
+### RPC Response Types
+
+```typescript
+interface EthBalanceResponse {
+  result: string; // Hex-encoded balance
+}
+```
+
+## 8. Implementation Requirements
+
+### Validation Rules
+
+1. **Address Validation**:
+
+   - All distributor addresses must be valid Ethereum addresses (checksum optional)
+   - Addresses normalized to checksummed format before queries
+
+2. **Date Validation**:
+
+   - Dates must be in ISO format (YYYY-MM-DD)
+   - Cannot query dates before distributor creation
+   - Cannot query future dates
+
+3. **Block Validation**:
+   - Block numbers must be positive integers
+   - Blocks must be at least 1000 blocks old (reorg protection)
+
+### Security Considerations
+
+1. **RPC Security**:
+
+   - Use HTTPS endpoints only
+   - Implement request rate limiting
+   - Handle RPC errors without exposing sensitive information
+
+2. **Data Integrity**:
+   - Atomic file writes to prevent corruption
+   - Validate all RPC responses before storage
+   - Maintain backup of existing data before updates
+
+## 9. Test Data Requirements
+
+### Minimum Test Data
+
+1. **Distributor List**: At least 3 distributor addresses with different creation dates
+2. **Block Numbers**: At least 7 days of block numbers spanning distributor lifetimes
+3. **RPC Responses**: Mock responses for `eth_getBalance` calls
+4. **Edge Cases**:
```

Comment:
Forget these lets just use all the distributors from our test data - the ones which are actual reward distributors